src/esw/imu_driver.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+import rospy
+from sensor_msgs.msg import Imu, MagneticField, Temperature
+from geometry_msgs.msg import Quaternion, Vector3
+from std_msgs.msg import Header
+from mrover.msg import CalibrationStatus
+
+import serial
+from serial import SerialException, SerialTimeoutException
+
+
+def main():
+    # publishers for all types of IMU data, queue size is 1 to make sure we don't publish old data
+    imu_pub = rospy.Publisher("imu/data", Imu, queue_size=1)
+    mag_pub = rospy.Publisher("imu/magnetometer", MagneticField, queue_size=1)
+    temp_pub = rospy.Publisher("imu/temp", Temperature, queue_size=1)
+    calibration_pub = rospy.Publisher("imu/calibration", CalibrationStatus, queue_size=1)
+
+    # no rate used because rate is effectively set by Arduino serial publisher
+    rospy.init_node("imu_driver")
+
+    # read serial connection info and IMU frame from parameter server
+    port = rospy.get_param("imu_driver/port", "/dev/imu")
+    baud = rospy.get_param("imu_driver/baud", 115200)
+    imu_frame = rospy.get_param("imu_driver/frame_id", "imu_link")
+
+    # create serial connection with Arduino
+    ser = serial.Serial(port, baud)
+    attempts = 0
+
+    while not rospy.is_shutdown():
+
+        # try to read a line from the serial connection,
+        # if this fails 100 times in a row then end the program
+        try:
+            line = ser.readline()
+            attempts = 0
+        except SerialException:
+            attempts += 1
+            if attempts > 100:
+                raise SerialTimeoutException(f"Unable to read from serial port {port}, timed out after 100 attempts")
```

Comment:
Hey does this mean that the entire program just crashes? Would you have to rerun roslaunch just to get the IMU program running again?

Reply:
Before it was fine because if a program died, it would automatically restart but now we don't do that

Reply:
well if this exception gets raised it will kill the node, but I think that's desired behavior since that would mean the IMU definitely isn't working or isn't plugged in.

since this will only happen either right after startup or when the IMU breaks/gets unplugged, I think it's fine to have to restart the roslaunch file, but if it's really necessary you can probably just rosrun this node separately to get it back online

Reply:
Oh that makes sense. Kinda sounds like a pain having to rosrun the node separately though if things somehow go bad

Reply:
well again, if things go bad there are very few circumstances when we wouldnt want to just restart the launch file right? Like in competition if this happens, then restarting the node almost certainly isnt gonna do any good, since theres something physically wrong with the IMU, so we would need to service it anyway.

Reply:
Yea makes sense